hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,656 @@
+# Another great side-effect is that all the templates can run locally as functions
+@wt.entrypoint
+@wt.dag
+def worker(worker_input: WorkerInput) -> WorkerOutput:
+    # We will need to use something like python-varname
+    # to automatically create a task called "setup_task" here.
+    setup_task = setup()
+    # setup_task will be a `Task` when building the workflow, vs an `hio.Output` object
+    # when running locally.
+
+
+    # Note how easy it is to reference variables from the DAG template input
+    # or previous tasks.
+    task_a = concat(ConcatInput(word_a=worker_input.value_a, word_b=setup_task.result))
+    task_b = concat(ConcatInput(word_a=worker_input.value_b, word_b=setup_task.result))
+    final_task = concat(ConcatInput(word_a=task_a.result, word_b=task_b.result))
+
+    # Note, we will automatically construct the dependency graph based on the
+    # input/output relationship between tasks, so Hera will infer the DAG as:
+    # setup_task >> [task_a, task_b] >> final_task
+
+    # Users will still be able to explicitly add dependencies between tasks
+    # using the rshift operator if they wish to define dependencies amongst
+    # tasks that don't share variables directly
+
+    # Easily "forward" task output parameters to the DAG's output parameters
+    # which replaces the existing syntax, which requires a "forward declaration"
+    # of the intended output parameter:
+    # with DAG(
+    #     ...,
+    #     outputs=Parameter(
+    #         name="my-dag-output",
+    #         value_from={"parameter": "{{task-name.outputs.parameters.param-name}}"},
+    #     ),
+    return WorkerOutput(value=final_task.result)
+```
+
+## Steps Template
+
+```python
+from hera.workflows import WorkflowTemplate
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the template's inputs
+class CalculatorInput(hio.Input):
+    x: int
+    y: int
+    operation: Literal["add", "sub"] = "add"
+
+@wt.script  # Adds a new script template to the workflow template called calculator
+def calculator(calc_input: CalculatorInput) -> hio.Output:
+    if calc_input.operation == "add":
+        return hio.Output(calc_input.x + calc_input.y)
+    return hio.Output(calc_input.x - calc_input.y)
+    
+
+# This defines another template's inputs
+class FiboInput(hio.Input):
+    num: int
+
+class FiboOutput(hio.Output):
+    num: int
+
+@wt.entrypoint
+@wt.steps  # Adds a new script template to the workflow template called fibonacci
+def fibonacci(fibo: FiboInput) -> FiboOutput:
+    # we will need to validate that users can only run/define certain kinds of expressions here
+    # and throw warnings if they are using syntax that is invalid
+    valid_number = hio.expr(fibo.num > 1)
+    previous_num = calculator(CalculatorInput(x=fibo.num, y=1, operation="sub"), when=valid_number)
+    previous_fibo = fibonacci(FiboInput(num=previous_num.result), when=valid_number)
```

Comment:
(discussion) How does this work if we run the Workflow on Argo? Is the idea that statically we figure out what the Workflow should look like? Or can we have an Argo Workflow with a dynamic number of nodes?

Reply:
It's static. The idea is that we do source code analysis to compile a supported subset of python to Argo.